[MachinePipeliner] Fix constraints aren't considered in certain cases

kasuga-fj opened the pull request:
when scheduling

When scheduling an instruction, if both any predecessors and any successors of the instruction are already scheduled, `SchedStart` isn't taken into account. It may result generating incorrect code. This patch fixes the problem. Also, this patch merges `SchedStart` into `EarlyStart` (same for `SchedEnd`).

I checked the schedule changes due to this patch in llvm-test-suite. In conclusion, as far as I observed, there was no change in most cases. Just in case, details of the results are as follows.

Compile options:
```
-O3 \
-mcpu=neoverse-v1 \
-mllvm -aarch64-enable-pipeliner \
-mllvm -pipeliner-max-stages=1000 \
-mllvm -pipeliner-max-mii=1000 \
-mllvm -pipeliner-enable-copytophi=0 \
-mllvm -pipeliner-ii-search-range=100 \
-mllvm -enable-misched=0 \
-mllvm -enable-post-misched=0
```

Observed changes:
| Test | Summary |
|-|-|
| MultiSource/Benchmarks/ASCI_Purple/SMG2000/smg2000 | The order of some instructions was changed |
| MultiSource/Benchmarks/TSVC/CrossingThresholds-flt/CrossingThresholds-flt | The order of some instructions was changed |
| MultiSource/Benchmarks/ASC_Sequoia/AMGmk/AMGmk | The order of some instructions was changed |

close #93936

kasuga-fj commented:
ping

bcahoon commented on llvm/lib/CodeGen/MachinePipeliner.cpp:
```diff
@@ -2420,47 +2420,48 @@ bool SwingSchedulerDAG::schedulePipeline(SMSchedule &Schedule) {
+        // backwards when all scheduled predecessors are loop-carried
+        // output/order dependencies. Empirically, there are also cases where
+        // scheduling becomes possible with backward search.
+        if (Schedule.onlyHasLoopCarriedOutputOrOrderPreds(SU, this)) {
```
I'm not sure I see how this keeps the original behavior?  I assume some test case fails without it?
If it's needed, can this be combined with the code at line 2455?

bcahoon replied:
I commented out the code between lines 2446-2449, and noticed some test failures. I took a quick look at sw-bad-sched.ll, and I can see what is going on.

Prior to this patch:
  ```
Inst (3)   %63:intregs = L2_loadri_io %15:intregs, 0 :: (load (s32) from %ir.lsr.iv1, !tbaa !0)
        es: 80000000 ls:        0 me: 7fffffff ms: ffffffff
  Trying to insert node between 0 and -1 II: 3
```

With this patch:
  ```
Inst (3)   %63:intregs = L2_loadri_io %15:intregs, 0 :: (load (s32) from %ir.lsr.iv1, !tbaa !0)
        es: ffffffff ls:        0
  Trying to insert node between 0 and -1 II: 3

```
In the old case, the pipeliner uses does a bottom up schedule because es == INT_MAX and ls == 0.
Because es is -1, with the patch, the pipeliner drops into the case when both es and ls have been defined. Previously, that used a top-down search (unless the SU is a Phi).

I think it's okay for the schedule processing to change after your patch. I feel it would be surprising if it didn't.  For example, in the case of swp-bad-sched.ll, the generated schedules are equivalent when comparing the output from your patch to the output from removing the code in lines 2446-2449. I haven't looked at the other cases though.

The question is - does the bottom-up scheduling produces a better result for certain situations. The patch has a comment "Empirically, there are also cases where scheduling becomes possible with backward search". This would be a reason to add/keep the code that does a backward search. But, I would see those cases and I think it would be better to combine any changes to do a bottom up search with the isPHI() check

kasuga-fj replied:
> does the bottom-up scheduling produces a better result for certain situations.
Yes. As far as I have checked with llvm-test-suite, there were cases where scheduling failed when stopping the backward search (about 1% of all cases). Removing lines 2446-2449 sometimes generated better schedule, but these were about a tenth of the cases where scheduling failed. Therefore I believe that keeping the original behavior is valuable.

In any case, as you mentioned, merging this process with isPHI() checks looks good to me. I'll fix it.

bcahoon reviewed (commented):
Can you add a test code using the example from the [issue](https://github.com/llvm/llvm-project/issues/93936)?

kasuga-fj commented:
Added test.

bcahoon reviewed (approved):
Thanks for fixing. LGTM.

kasuga-fj committed:
[MachinePipeliner] Fix constraints aren't considered in certain cases
when scheduling

When scheduling an instruction, if both any predecessors and any
successors of the instruction are already scheduled, `SchedStart` isn't
taken into account. It may result generating incorrect code. This patch
fixes the problem.
